### ISSUE TITLE
fails to link `curl::init()`'s call to `INIT_CTOR` on macOS 12.0 (Monterey)

### ISSUE THREAD

**Issue by cormacrelf** [alexcrichton/curl-rust#417]
### Summary

Only on macOS. May occur on macOS earlier than 12 using Xcode 13, I'm not sure. Underlying issue is https://github.com/rust-lang/rust/issues/90342

You get a linker error, there is a workaround for end users of the crate (see the rust issue; set the env var). I think curl-rust specifically can avoid this problem.

### Repro:

- rust 1.56 stable
- macOS 12.0.1 Monterey
-  `cargo test` in the curl-rust repo (currently at df64ee4).

This is using either system libcurl or with `--features static-curl`. Makes no difference.

```
$ cargo test

... very long cc -arch arm64 invocation
...
  = note: ld: reference to symbol (which has not been assigned an address) __ZN4curl4init9INIT_CTOR17h97cc33cf050cb462E in '__ZN4curl4init17ha644d831c2a57f65E' from /Users/cormac/git/tryout/libcurl-monterey/target/debug/deps/libcurl-0f9cbb7dde66dd88.rlib(curl-0f9cbb7dde66dd88.curl.0b6dcf6e-cgu.2.rcgu.o) for architecture arm64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### Possible solutions, other than waiting for the rust issue

The relevant symbol is defined at:

https://github.com/alexcrichton/curl-rust/blob/df64ee4fcb1e3e2927d04e6bb6d8a1c0feedc844/src/lib.rs#L93-L101

I think the actual bug is due to the the workaround for https://github.com/rust-lang/rust/issues/47384 referenced further down, in which the `init` function calls `INIT_CTOR();`.

I think you can solve it by:

- Adding `#[used]` to `INIT_CTOR` (any MSRV issue? not sure)
- Moving `INIT_CTOR` outside the `init` function
- Change `pub fn init() { INIT_CTOR(); }` to `pub fn init() { init_inner(); }`

It works on macOS 12, at least. But I think it would be good enough, because:

- We're just trying to get rust to preserve `INIT_CTOR`, despite that rust-lang/rust issue 47384 where not using any code from the module means it won't get linked at all
- Does curl-rust have a problem with people using the library but managing not to use any code from it? If no code is used, do we need the pre-main initializer to run at all?
- I don't think so. It's not obvious to me that curl-rust is actually affected by the thing it has a workaround for.

**Comment by shepmaster** [comment]
Some [playground utilities use cargo](https://github.com/integer32llc/rust-playground/blob/4f64f6b80584a8b6499e475f24bb0bcde2200085/top-crates/Cargo.toml#L12) which uses curl. I ran into this problem.

As a workaround that doesn't involve updating any of my dependencies, I can make use of the [`MACOSX_DEPLOYMENT_TARGET` idea](https://github.com/alexcrichton/curl-rust/issues/417#issuecomment-953059447):

```
% cargo clean
% MACOSX_DEPLOYMENT_TARGET=11 cargo run
```

This allows me to hobble on until it's properly fixed by y'all wonderful people.

**Comment by sagebind** [comment]
Can someone running macOS Monterey verify that the latest `main` now compiles without any extra environment variables? #426 _should_ fix the issue, but I don't have such a device to verify myself.

**Comment by shepmaster** [comment]
```
% git show --pretty=oneline --stat
0aea09c428b9bc2bcf46da0fc33959fe3f03c74a (HEAD -> main, origin/main, origin/HEAD) Refactor init to avoid link bugs on macOS Monterey (#426)
 src/lib.rs | 102 ++++++++++++++++++++++++++++++++++++++++---------------------
 1 file changed, 68 insertions(+), 34 deletions(-)

% cargo build
    Updating crates.io index
   Compiling cc v1.0.72
   Compiling pkg-config v0.3.22
   Compiling libc v0.2.107
   Compiling curl v0.4.40 (/Users/shep/Projects/curl-rust)
   Compiling libz-sys v1.1.3
   Compiling curl-sys v0.4.50+curl-7.80.0 (/Users/shep/Projects/curl-rust/curl-sys)
   Compiling socket2 v0.4.2
    Finished dev [unoptimized + debuginfo] target(s) in 4.12s
curl-rust %
```

**Comment by sagebind** [comment]
Nice! We'll get this fix out to Crates.io soon.

**Comment by sagebind** [comment]
The fix for this is now available in curl 0.4.41.
